controllers/accounts.go

```diff
@@ -256,45 +324,227 @@ func (ac *AccountsController) SetupPasswordFinalize(w http.ResponseWriter, r *ht
 		return
 	}
 
-	account, err := ac.opaqueService.SetupPasswordFinalize(verification.Email, opaqueRecord)
+	registrationState, err := ac.opaqueService.SetupPasswordFinalize(verification.Email, opaqueRecord)
 	if err != nil {
 		switch {
-		case errors.Is(err, util.ErrRegistrationStateNotFound):
+		case errors.Is(err, util.ErrInterimPasswordStateNotFound):
 			util.RenderErrorResponse(w, r, http.StatusNotFound, err)
-		case errors.Is(err, util.ErrRegistrationStateExpired):
+		case errors.Is(err, util.ErrInterimPasswordStateExpired):
 			util.RenderErrorResponse(w, r, http.StatusBadRequest, err)
 		default:
 			util.RenderErrorResponse(w, r, http.StatusInternalServerError, err)
 		}
 		return
 	}
 
-	if err = ac.ds.UpdateAccountLastEmailVerifiedAt(account.ID); err != nil {
+	if err = ac.ds.UpdateAccountLastEmailVerifiedAt(*registrationState.AccountID); err != nil {
+		util.RenderErrorResponse(w, r, http.StatusInternalServerError, err)
+		return
+	}
+
+	var authToken string
+	if !registrationState.RequiresTwoFA {
+		authToken, err = ac.createSessionAfterPasswordSetup(*registrationState.AccountID, r.UserAgent(), verification.ID)
+		if err != nil {
+			util.RenderErrorResponse(w, r, http.StatusInternalServerError, err)
+			return
+		}
+	}
+
+	render.Status(r, http.StatusOK)
+	render.JSON(w, r, &PasswordFinalizeResponse{
+		AuthToken:     authToken,
+		RequiresTwoFA: registrationState.RequiresTwoFA,
+	})
+}
+
+// @Summary Finalize password setup with 2FA
+// @Description Complete the password setup process after 2FA verification. If a recovery key is used, 2FA will be disabled.
+// @Tags Accounts
+// @Accept json
+// @Produce json
+// @Param Authorization header string true "Bearer + verification token"
+// @Param Brave-Key header string false "Brave services key (if one is configured)"
+// @Param request body services.TwoFAAuthRequest true "2FA verification request"
+// @Success 200 {object} LoginFinalize2FAResponse
+// @Failure 400 {object} util.ErrorResponse
+// @Failure 401 {object} util.ErrorResponse
```

Comment:
We should probably mention that this is the status code that will be used if the user types the wrong TOTP code.

Reply:
We also return HTTP 401 for other errors that may not be related to user error. A client should probably rely on the codes defined in error.go for error variant identification.